### PR TITLE
architecture: core: add a section on registration of shared memory

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Adds a section with a few diagrams describing how memory is shared with
secure world both with SMC and FF-A based communication.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

### CODE REVIEW

File: architecture/core.rst [hunk 1]
```diff
@@ -1319,6 +1319,123 @@ of shared memory physical chunks. OP-TEE linux kernel driver relies on linux
 kernel dma-buf support (``CONFIG_DMA_SHARED_BUFFER``) to track shared memory
 buffers references.
 
+Registering shared memory
+=========================
+
+Only dynamic or physically non-contiguos shared memory needs to be
```
Comment (jbech-linaro) [inline architecture/core.rst:1325]: s/contiguos/contiguous/g

File: architecture/core.rst [hunk 2]
```diff
@@ -1319,6 +1319,123 @@ of shared memory physical chunks. OP-TEE linux kernel driver relies on linux
...
+   return is combined with an implicit registration of shared memory. The
+   registration is ended with a ``OPTEE_SMC_RETURN_RPC_FREE`` request.
+
+.. uml::
+    :align: center
+    :caption: Register shared memory example
+
+    participant "Normal World\nOS Kernel" as ns
+    participant "Secure World\nOP-TEE OS" as sec
+
```
Comment (jbech-linaro) [inline architecture/core.rst:1349]: Unnecessary spaces.

File: architecture/core.rst [hunk 3]
```diff
@@ -1319,6 +1319,123 @@ of shared memory physical chunks. OP-TEE linux kernel driver relies on linux
...
+    sec -> sec  : Register shared memory passed 
+    sec -> ns : Return
+
+.. uml::
+    :align: center
+    :caption: Unregister shared memory example
+
+    participant "Normal World\nOS Kernel" as ns
+    participant "Secure World\nOP-TEE OS" as sec
+
```
Comment (jbech-linaro) [inline architecture/core.rst:1360]: Unnecessary spaces.

File: architecture/core.rst [hunk 4]
```diff
@@ -1319,6 +1319,123 @@ of shared memory physical chunks. OP-TEE linux kernel driver relies on linux
...
+SMC based OP-TEE MSG ABI
+------------------------
+
+With the SMC based OP-TEE MSG ABI there are a few exceptions where memory
+doesn't need to be shared before it can be accessed from OP-TEE OS. These
+are:
+
+1. When issuing the SMC ``OPTEE_SMC_CALL_WITH_ARG`` where the physical
+   address of the supplied ``struct optee_msg_arg`` is passed in one register.
```
Comment (jbech-linaro) [inline architecture/core.rst:1337]: Just a suggestion:
s/one register/as one of the registers/

File: architecture/core.rst [hunk 5]
```diff
@@ -1319,6 +1319,123 @@ of shared memory physical chunks. OP-TEE linux kernel driver relies on linux
...
+    alt cookie not known
+        sec -> spmc : FFA_MEM_RETRIEVE_REQ(cookie)
+        spmc -> sec : Return memory description
+        sec -> sec : Register shared memory
+    end
+    sec -> sec : Process the yielding call
+    sec -> ns : Return
+
+Unregistration of shared memory is also done in two steps. First with a
+call to OP-TEE and then with a call to the SPMC. If the lazy  second
```
Comment (jbech-linaro) [inline architecture/core.rst:1415]: double spaces

**Comment by jbech-linaro** [comment]
Thanks Jens, just minor stuff. With those addressed, feel free to use my R-B.
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

**Comment by jenswi-linaro** [comment]
Thanks, tag applied

File: architecture/core.rst [hunk 6]
```diff
@@ -1319,6 +1319,124 @@ of shared memory physical chunks. OP-TEE linux kernel driver relies on linux
 kernel dma-buf support (``CONFIG_DMA_SHARED_BUFFER``) to track shared memory
 buffers references.
 
+Registering shared memory
+=========================
+
+Only dynamic or physically non-contiguous shared memory needs to be
+registered. Static or physically contiguos shared memory is already known
```
Comment (jbech-linaro) [inline architecture/core.rst:1326]: There is one more "`contiguos`" left.

Comment (jenswi-linaro) [reply architecture/core.rst]: Sorry, should be fixed now.

**Pull request merged**